lib/charms/ovn_charm.py

```diff
@@ -188,6 +188,34 @@ def __init__(self, **kwargs):
                 self.restart_map.update({
                     '/etc/tmpfiles.d/nova-ovs-vhost-user.conf': []})
 
+    def is_port(self, port, status_set=True):
+        """Check if the port exists.
+
+        :param port: interface name
+        :type port: str
+        :param status_set: change the juju state to blocked if
+                           the port does not exist
+        :type status_set: bool
+        :returns: boolean whether port exists
+        :rtype: bool
+        :raises: ValueError
+        """
+        ch_core.hookenv.log('checking port "{}" existence'.format(port),
+                            level=ch_core.hookenv.DEBUG)
+        try:
+            if self.options.enable_dpdk:
+                # TODO: add function to check DPDK interface
+                pass
+            else:
+                self.run('ip', 'link', 'show', port)
```

Comment:
I am a bit concerned about this approach, are you sure you will detect all possible failure modes here?

I would be more comfortable with adding a try/except clause around this part:
https://github.com/openstack-charmers/charm-layer-ovn/blob/e987e12ec92a00d630ddc0eaf5823001b7429910/lib/charms/ovn_charm.py#L589-L600

That would also free you from having to deal with intricate details of different plugging mechanisms.

Reply:
If I use try / except, I will end up again with an error when trying to set the no valid interface down.
Example:
```bash
juju config ovn-chassis bridge-interface-mappings = "br-ex: no-valid-port"
# ends up in a blocked state
juju config ovn-chassis bridge-interface-mappings = "br-ex: valid-port"
# ended in an error state while trying to set the port down and needs to be resolved
```

Reply:
Do you think it would be good to omit deleting in error?

Reply:
Interesting, what if we detect the error and remove the port without doing linkdown in the error-handler?

I see that `ovs-vsctl` behaves a bit strange here, example:
```
root@ovs:~# ovs-vsctl add-port testbr non-existing-port;echo $?
ovs-vsctl: Error detected while setting up 'non-existing-port': could not open network device non-existing-port (No such device).  See ovs-vswitchd log for details.
ovs-vsctl: The default log directory is "/var/log/openvswitch".
0
root@ovs:~# ovs-vsctl show
9e63364b-aaa1-4e7d-8fc7-e801276ea2b8
    Bridge testbr
        Port non-existing-port
            Interface non-existing-port
        Port testbr
            Interface testbr
                type: internal
    ovs_version: "2.13.0"
```
I.e. it does not raise an error code but only prints the error, and also leaves the port in the configuration. The charm could possibly detect that though by parsing the output from `ovs-vsctl`.

If error detected then do a `del_bridge_port(port, linkdown=False)` immediately.

Then on next configuration change you will not attempt to remove the port left behind from a previous execution.

What do you think?

Reply:
This sounds like a good idea, but I can't imagine how to reasonably detect a error. If I'm not mistaken, I can't read the result of the subproccess.check_call after it was called? Or do you mean to read it from an oats-vsctl show?

Reply:
You can examine the output from the command with `subprocess.check_output`.

An alternative could also be to add error checking around the call to `del_bridge_port`, if it fails rerun it with `linkdown=False` for example? The `ovs-vsctl` command is called with `--if-exists` so it will not fail when the port does not exist.

Reply:
The result of my proposal looks like this.
There are two other questions:
- Is ʻexcept subprocess.CalledProcessError` enough? (as mentioned by @coreycb)
- Do we need an error log because there is an alert log?

![bug1896913-status-blocked](https://user-images.githubusercontent.com/34167657/96557405-833d9400-12ba-11eb-8a33-5d0a9d84a44f.png)

Reply:
While doing functional tests for these changes, I ran into a problem with `charm relations` and decided to use` custom_assess_status_last_check`.

![bug18969914-juju-status-blocked](https://user-images.githubusercontent.com/34167657/96865451-43b5aa00-146a-11eb-96f6-d6deed35a42f.png)